std/bitmanip.d

```diff
@@ -1710,6 +1710,242 @@ public:
+
+    unittest
+    {
+        static if (size_t.sizeof == 8)
+        {
+            size_t x = 0x12345678_90ABCDEF;
+            size_t y = 0xFEDBCA09_87654321;
+
+            assert(rollLeft(x, y, 32) == 0x90ABCDEF_FEDBCA09);
+            assert(rollLeft(y, x, 4) == 0xEDBCA098_76543211);
+        }
+        else static if (size_t.sizeof == 4)
+        {
+            size_t x = 0x12345678;
+            size_t y = 0x90ABCDEF;
+
+            assert(rollLeft(x, y, 16) == 0x567890AB);
+            assert(rollLeft(y, x, 4) == 0x0ABCDEF1);
+        }
+    }
+
+    /**
+     * Operator $(D >>=) support.
+     *
+     * Shifts all the bits in the array to the right by the given number of
+     * bits.  The rightmost bits are dropped, and 0's are inserted at the front
+     * to fill up the vacant bits.
+     *
+     * $(RED Warning: unused bits in the final word up to the next word
+     * boundary may be overwritten by this operation. It does not attempt to
+     * preserve bits past the end of the array.)
+     */
+    void opOpAssign(string op)(size_t nbits) @nogc pure nothrow
+        if (op == ">>")
+    {
+        size_t wordsToShift = nbits / bitsPerSizeT;
+        size_t bitsToShift = nbits % bitsPerSizeT;
+
+        if (wordsToShift < dim)
+        {
+            // Unfortunately, due to the way core.bitop.bt() works, indices are
+            // the reverse order of what an equivalent >> on a size_t would do,
+            // so we have to use rollLeft() instead of rollRight() here,
+            // contrary to the direction of the >> operator on this BitArray.
+            foreach_reverse (i; 1 .. dim - wordsToShift)
+            {
+                ptr[i + wordsToShift] = rollLeft(ptr[i], ptr[i-1],
+                                                 bitsToShift);
+            }
+            ptr[wordsToShift] = rollLeft(ptr[0], 0, bitsToShift);
+        }
+
+        import std.algorithm : min;
+        foreach (i; 0 .. min(wordsToShift, dim))
+        {
+            ptr[i] = 0;
+        }
+    }
+
+    /**
+     * Operator $(D <<=) support.
+     *
+     * Shifts all the bits in the array to the left by the given number of
+     * bits.  The leftmost bits are dropped, and 0's are appended to the end to
+     * fill up the vacant bits.
+     *
+     * $(RED Warning: unused bits in the final word up to the next word
+     * boundary may be overwritten by this operation. It does not attempt to
+     * preserve bits past the end of the array.)
+     */
+    void opOpAssign(string op)(size_t nbits) @nogc pure nothrow
+        if (op == "<<")
+    {
+        size_t wordsToShift = nbits / bitsPerSizeT;
+        size_t bitsToShift = nbits % bitsPerSizeT;
+
+        // Unfortunately, due to the way core.bitop.bt() works, indices are the
+        // reverse order of what an equivalent << on a size_t would do, so we
+        // have to use rollRight() instead of rollLeft() here, contrary to the
+        // direction of the << operator on this BitArray.
```

Comment:
Ah, I see what's going on, you're shifting in the wrong directions.
In fact the string formatting is LSB first, so shifting in one direction will shift the string into the other.

```d
import std.bitmanip, std.format;

void main()
{
    BitArray ba;
    ba.init([0, 0, 0, 0, 1]);
    assert(ba[4]);
    assert(format("%b", ba) == "00001");
    ba >>= 2;
    assert(ba[2]);
    assert(format("%b", ba) == "001");
    ba <<= 2;
    assert(ba[4]);
    assert(format("%b", ba) == "00001");
}
```